File: slnSopimusrekisteriLuokat/Sopimusrekisteri.BLL_CF/Tiedosto.cs
using System;

namespace Sopimusrekisteri.BLL_CF
{
    public class Tiedosto : IEditable
  {
		#region Propertyt

		public int Id {get; private set;}
		public string TiedostoNimi {get; set;}
		public string URL {get; set;}
		public string Selite {get; set;}
		public int? TiedostoLahdeId {get; set;}
		public int? SopimusId {get; set;}
		public int? SharePointId {get; set;}
		public int? ArkistonSijaintiId {get; set;}
		public string DocumentID {get; set;}
		public string ArkistointiTunniste {get; set;}
		public int? Sivuja {get; set;}
		public string Info {get; set;}
		public DateTime Luotu {get; set;}
		public string Luoja {get; set;}
		public DateTime? Paivitetty {get; set;}
		public string Paivittaja {get; set;}
		public int? SopimusFormaattiId {get; set;}
		public string Asiakirjatarkenne {get; set; }
		public Guid? MFilesVault { get; set; }
		public int MFilesType { get; set; }
		public int? MFilesId { get; set; }
		public Guid? MFilesObject { get; set; }

		#endregion

		#region Read-only

		public string MFilesLink => GetMFilesLink(MFilesAction.Show);
		public TiedostonSijainti Sijainti => !string.IsNullOrEmpty(MFilesLink) ? TiedostonSijainti.MFiles : TiedostonSijainti.Sharepoint;

        #endregion

        #region Entiteettiviittaukset

        public virtual Tiedostolahde TiedostoLahde {get; set;}
		public virtual Sopimus Sopimus {get; set;}
		public virtual ArkistonSijainti ArkistonSijainti {get; set;}
		public virtual SopimusFormaatti SopimusFormaatti {get; set;}

		#endregion

		#region Metodit

		public override bool Equals(object obj)
		{
			var tmp = obj as Tiedosto;
			if (tmp == null) return false;
			return tmp.Id == Id;
		}

		public override int GetHashCode()
		{
			return Id.GetHashCode();
		}

		public override string ToString()
		{
			return Id.ToString();
		}

		public string GetMFilesLink(MFilesAction action)
        {
			if (MFilesVault.HasValue && MFilesId.HasValue)
            {
				string actionStr = action.ToString().ToLower();

				string link = $"m-files://{actionStr}/{MFilesVault.ToString().ToUpper()}/{MFilesType}-{MFilesId}";

				if (MFilesObject.HasValue)
					link += $"?object={MFilesObject.ToString().ToUpper()}";

				return link;
			}

			return string.Empty;
        }

		#endregion

  }
}

File: slnSopimusrekisteri/appSopimusrekisteri.DTO/Tiedosto.vb
Public Class Tiedosto

    Public Property Id As Integer?
    Public Property Nimi As String
    Public Property Url As String
    Public Property Selite As String
    Public Property LahdeId As Integer?
    Public Property Lahde As String
    Public Property SopimusId As Integer?
    Public Property SharepointId As Integer?
    Public Property ArkistonSijaintiId As Integer?
    Public Property ArkistonSijainti As String
    Public Property DocumentId As String
    Public Property ArkistointiTunniste As String
    Public Property Sivuja As Integer?
    Public Property Info As String
    Public Property Luoja As String
    Public Property Luotu As Date?
    Public Property Paivittaja As String
    Public Property Paivitetty As Date?
    Public Property AsiakirjaTarkenneId As Integer?
    Public Property AsiakirjaTarkenne As String
    Public Property SopimusFormaattiId As Integer?
    Public Property SopimusFormaatti As String
    Public Property MFilesVault As Guid?
    Public Property MFilesType As Integer
    Public Property MFilesId As Integer?
    Public Property MFilesObject As Guid?

    Public ReadOnly Property MFilesLink As String
        Get
            Return GetMFilesLink(Enumeraattorit.MFilesAction.Show)
        End Get
    End Property

    Public ReadOnly Property Sijainti As Enumeraattorit.TiedostonSijainti
        Get
            Return If(Not String.IsNullOrEmpty(MFilesLink), Enumeraattorit.TiedostonSijainti.MFiles, Enumeraattorit.TiedostonSijainti.Sharepoint)
        End Get
    End Property

    Public Function GetMFilesLink(ByVal action As Enumeraattorit.MFilesAction) As String
        If MFilesVault.HasValue AndAlso MFilesId.HasValue Then
            Dim actionStr As String = action.ToString().ToLower()
            Dim link As String = $"m-files://{actionStr}/{MFilesVault.ToString().ToUpper()}/{MFilesType}-{MFilesId}"
            If MFilesObject.HasValue Then link += $"?object={MFilesObject.ToString().ToUpper()}"
            Return link
        End If

        Return String.Empty
    End Function


End Class
